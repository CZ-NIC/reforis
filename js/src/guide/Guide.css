/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

.wrapper-guide {
    flex-direction: column;
}

.wrapper-guide #content {
    display: flex;
    align-items: center;
    padding: 2em;
}

#stepsbar,
#guide-nav-container {
    flex: 0.2;
    width: 100%;
    background: #f2f2f2;
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
    align-items: center;
    justify-content: center;
}

#stepsbar-header {
    display: flex;
    flex-direction: column;
    align-items: flex-end;
    padding: 1em;
    background: #f2f2f2;
    white-space: nowrap;
}

#stepsbar ul {
    display: flex;
    flex-direction: row;
    align-items: center;
    flex-wrap: wrap;
}

#stepsbar ul li a {
    padding: 0.5rem 2rem;
    font-size: 1.1em;
    display: block;
    color: #787878;

    white-space: nowrap;
    flex-basis: 100%;
}

#stepsbar ul li a:hover {
    background: #cbe8f5;
    text-decoration: none;
}

#stepsbar li a .fas {
    margin-right: 0.5rem;
}

#stepsbar a.active {
    color: #fff;
    background: #00a2e2;
}

#stepsbar a.passed {
    color: #000;
}

#stepsbar a.passed.active {
    color: #fff;
    background: #00a2e2;
}

#stepsbar a.next {
    color: #00a2e2;
}

#stepsbar ul a.next.active {
    color: #fff;
    background: #00a2e2;
}

@media screen and (max-width: 575px) {
    #stepsbar ul {
        flex-direction: column;
    }
}

#workflow-selector {
    display: flex;
    flex-direction: row;
    justify-items: center;
    align-content: center;
    flex-wrap: wrap;
}

#workflow-selector .btn {
    margin: 1rem;
}

.workflow {
    display: flex;
    flex-direction: column;
    align-items: center;
}

.workflow {
    width: 100%;
}

#guide-page-container {
    display: flex;
    flex-direction: row;
    align-items: center;
}

#guide-page {
    max-width: 100%;
    flex-grow: 2;
}

.blinking {
    background-color: #00a2e2;
    color: white;
    -webkit-animation-name: blinking;
    -webkit-animation-duration: 1.8s;
    -webkit-animation-iteration-count: infinite;
    animation-name: blinking;
    animation-duration: 1.8s;
    animation-iteration-count: infinite;
}

.guide-card {
    background-color: #cbe8f5;
}

.guide-card p:last-of-type {
    margin-bottom: 0;
}

/* Safari 4.0 - 8.0 */
@-webkit-keyframes blinking {
    0% {
        opacity: 1;
    }
    50% {
        opacity: 0.2;
    }
    100% {
        opacity: 1;
    }
}

/* Guide controls */
.sticky-top {
    top: 0.5em;
}
