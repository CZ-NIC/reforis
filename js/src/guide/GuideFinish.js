/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import { Button, formFieldsSize } from "foris";

import useGuideFinish from "./hooks";

export default function GuideFinished() {
    const onGuideFinishHandler = useGuideFinish();
    return (
        <>
            <h2>{_("Guide Finished")}</h2>
            <p>{_("Once you leave this guide you'll be granted access to the full configuration interface of this device.")}</p>
            <p>{_("To further improve your security consider installing data collection plugin (via Package Management). This will allow you to be part of our security research to discover new attackers and give you access to dynamic firewall updates blocking known attackers.")}</p>
            <div className={`${formFieldsSize} text-right`}>
                <Button forisFormSize onClick={onGuideFinishHandler}>{_("Continue")}</Button>
            </div>
        </>
    );
}
