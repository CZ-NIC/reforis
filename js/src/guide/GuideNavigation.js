/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { withRouter } from "react-router";
import { Link, NavLink } from "react-router-dom";
import PropTypes from "prop-types";

import useGuideFinish from "./hooks";
import STEPS from "./steps";

GuideNavigation.propTypes = {
    workflow_steps: PropTypes.arrayOf(PropTypes.string).isRequired,
    passed: PropTypes.arrayOf(PropTypes.string).isRequired,
    next_step: PropTypes.string.isRequired,
};

export default function GuideNavigation({ workflow_steps, passed, next_step }) {
    const onGuideFinishHandler = useGuideFinish();
    const navigationItems = workflow_steps.map(
        (step) => (
            <GuideNavigationItem
                key={step}
                name={STEPS[step].name}
                passed={passed.includes(step)}
                url={`/${step}`}
                next={step === next_step}
            />
        ),
    );

    return (
        <>
            <ul className="list-unstyled">
                {navigationItems}
            </ul>
            <NextStepButtonWithRouter next_step={next_step} />
            <button type="button" className="btn btn-link" onClick={onGuideFinishHandler}>
                {_("Skip guide")}
            </button>
        </>
    );
}

GuideNavigationItem.propTypes = {
    name: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
    passed: PropTypes.bool.isRequired,
    next: PropTypes.bool.isRequired,
};

function GuideNavigationItem({
    name, url, next, passed,
}) {
    const passedClassName = passed ? "passed" : "";
    const nextClassName = next ? "next" : "";

    const content = (
        <>
            <i className="fas fa-arrow-right" />
            {name}
        </>
    );

    return (
        <li>
            {passed || next
                ? (
                    <NavLink className={`${passedClassName} ${nextClassName}`} to={url}>
                        {content}
                    </NavLink>
                )
                // eslint-disable-next-line jsx-a11y/anchor-is-valid
                : <a>{content}</a>}
        </li>
    );
}

NextStepButton.propTypes = {
    next_step: PropTypes.string.isRequired,
    location: PropTypes.object.isRequired,
};

function NextStepButton({ next_step, location }) {
    if (location.pathname === `/${next_step}`) return null;
    return (
        <Link id="next-step-button" className="btn btn-lg btn-light " to={`/${next_step}`}>
            <i className="fas fa-arrow-right" />
        </Link>
    );
}

const NextStepButtonWithRouter = withRouter(NextStepButton);
