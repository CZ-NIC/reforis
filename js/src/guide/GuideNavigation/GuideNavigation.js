/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import STEPS from "../steps";
import GuideNavigationItem from "./GuideNavigationItem";

GuideNavigation.propTypes = {
    workflow_steps: PropTypes.arrayOf(PropTypes.string).isRequired,
    passed: PropTypes.arrayOf(PropTypes.string).isRequired,
    next_step: PropTypes.string.isRequired,
};

export default function GuideNavigation({ workflow_steps, passed, next_step }) {
    const navigationItems = workflow_steps.map((step) => (
        <GuideNavigationItem
            key={step}
            name={STEPS[step].name}
            passed={passed.includes(step)}
            url={`/${step}`}
            next={step === next_step}
        />
    ));

    return (
        <>
            <ul className="list-unstyled">{navigationItems}</ul>
        </>
    );
}
