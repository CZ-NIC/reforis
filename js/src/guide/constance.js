/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import Password from '../password/Password';
import WorkflowSelect from './WorkflowSelect';
import Interfaces from '../interfaces/Interfaces';
import RegionAndTime from '../regionAndTime/RegionAndTime';
import DNS from '../dns/DNS';
import Updates from '../updates/Updates';
import WAN from '../wan/WAN';
import LAN from '../lan/LAN';
import GuideFinished from './GuideFinish';

export const URL_PREFIX = '/guide';

export const STEPS = {
    password: {
        name: _('Password'),
        component: Password
    },
    profile: {
        name: _('Workflow'),
        component: WorkflowSelect
    },
    networks: {
        name: _('Interfaces'),
        component: Interfaces
    },
    time: {
        name: _('Time'),
        component: RegionAndTime
    },
    dns: {
        name: _('DNS'),
        component: DNS
    },
    updater: {
        name: _('Updates'),
        component: Updates
    },
    wan: {
        name: _('WAN'),
        component: WAN
    },
    lan: {
        name: _('LAN'),
        component: LAN
    },
    finished: {
        name: _('Finish'),
        component: GuideFinished
    }
};
