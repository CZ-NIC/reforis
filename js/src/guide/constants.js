/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

export const GUIDE_URL_PREFIX = "/guide";

const HELP_CONTENT_DEFAULTS = {
    password: {
        initial: [_("Welcome to Foris web interface. This guide will help you to setup your device. Firstly it is required to set your password. Note the security of your home network is in your hands, so try not to use weak passwords.")],
        completed: _("Your password is set. You may proceed to the next step."),
    },
    profile: {
        completed: _("The workflow was set. You may proceed to next step."),
    },
    networks: {
        initial: [
            _("Here you need to decide which interfaces belong to which network."),
            _("If you are in doubt use the current settings."),
        ],
        completed: _("You've configured your network interfaces. It seems that you didn't break any crucial network settings so you can safely proceed to the next step."),
    },
    time: {
        initial: _("Now you need to set the time and timezone of your device."),
        completed: _("Your time and timezone seem to be set. Please make sure that the time matches the time on your computer if not try to update it via NTP or manually."),
    },
    dns: {
        initial: _("A proper DNS resolving is one of the key security features of your device. Let's configure it now."),
        completed: _("You've updated your DNS configuration. Try to run connection test to see whether your DNS resolver is properly set."),
    },
    wan: {
        initial: _("In order to access the internet, you need to configure your WAN interface."),
        completed: _("You've configured your WAN interface. Try to run connection test to see whether it is working properly and if so you can safely proceed to the next step."),
    },
    lan: {
        initial: _("Now you should configure your LAN interface. Note that when you change your network settings you probably won't be able to connect to the configuration interface unless you restart the network on your current device."),
        completed: _("You've configured your LAN interface. Try to test whether settings work properly and if so you can safely proceed to the next step."),
    },
};

export const HELP_CONTENT = {
    unset: { ...HELP_CONTENT_DEFAULTS },
    router: {
        ...HELP_CONTENT_DEFAULTS,
        finished: {
            initial: _("Congratulations! Now your device is able to act as a router and route network traffic among LAN, WAN and guest network."),
        },
    },
    min: {
        ...HELP_CONTENT_DEFAULTS,
        finished: {
            initial: _("Minimal device setup is finished. Note that you probably need to perform some further configuration updates to fit the device to your needs."),
        },
    },
    bridge: {
        ...HELP_CONTENT_DEFAULTS,
        networks: {
            ...HELP_CONTENT_DEFAULTS.networks,
            initial: [
                HELP_CONTENT_DEFAULTS.networks.initial[0],
                _("You chose to act as local server this means that it doesn't make sense to put any interfaces to your WAN and Guest Network. So it is a good idea to assign all interfaces to LAN."),
            ],
        },
        lan: {
            initial: [
                _("To act as a local server, there's no need to manage LAN (if you still want to manage it reset the guide and choose the Router workflow). You probably want to act as a client here thus select <strong>Computer</strong> mode here."),
                _("If you select the <strong>static</strong> configuration be sure that the IP addresses are entered correctly otherwise you won't be able to access this configuration interface when the new settings are applied."),
                _("If you select the <strong>DHCP</strong> method you probably you need to obtain a new IP address of this device. You can obtain it from the DHCP server which is managing your LAN. Then you need to connect to the new IP address to proceed the guide."),
                _("Note that either way you might need to re-plug your ethernet cables after you update the settings here."),
            ],
            completed: _("Please test whether the settings you provided are correctly working. This means that you can access the configuration interface of your device and your device is able to access the internet (you can use the connection test below)."),
        },
        finished: {
            initial: _("Congratulations! Now your device should be able to act as a server on your local network."),
        },
    },
};
