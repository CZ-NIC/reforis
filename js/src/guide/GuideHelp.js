import React from "react";
import PropTypes from "prop-types";

import { HELP_CONTENT } from "./constants";

GuideHelp.propTypes = {
    workflow: PropTypes.string.isRequired,
    step: PropTypes.string.isRequired,
    completed: PropTypes.bool,
};

GuideHelp.defaultProps = {
    completed: false,
};

export default function GuideHelp({ workflow, step, completed }) {
    const stepContent = HELP_CONTENT[workflow][step];
    if (!stepContent) {
        return null;
    }

    const isCompletedVisible = stepContent && completed && stepContent.completed;
    if (!(stepContent.initial || isCompletedVisible)) {
        return null;
    }

    return (
        <div className="card guide-card my-4">
            <div className="card-body">
                {stepContent.initial && <ParagraphsArray content={stepContent.initial} />}
                {isCompletedVisible && <p className="font-weight-bold">{ stepContent.completed }</p>}
            </div>
        </div>
    );
}

ParagraphsArray.propTypes = {
    content: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.arrayOf(PropTypes.string),
    ]),
};

function ParagraphsArray({ content }) {
    return Array.isArray(content)
        ? content.map((line, index) => <p key={index} dangerouslySetInnerHTML={{ __html: line }} />)
        : <p dangerouslySetInnerHTML={{ __html: content }} />;
}
