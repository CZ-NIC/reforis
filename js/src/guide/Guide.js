/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import { BrowserRouter } from "react-router-dom";
import { Redirect, Route, Switch } from "react-router";
import PropTypes from "prop-types";

import {
    Portal, REFORIS_URL_PREFIX, useAPIGet, withErrorMessage, withSpinnerOnSending,
    AlertContextProvider,
} from "foris";

import API_URLs from "common/API";

import { GUIDE_URL_PREFIX } from "./constants";
import GuideNavigation from "./GuideNavigation";
import GuideHelp from "./GuideHelp";
import STEPS from "./steps";

Guide.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function Guide({ ws }) {
    const [guideData, getGuideData] = useAPIGet(API_URLs.guide);
    useEffect(() => {
        getGuideData();
    }, [getGuideData]);

    return (
        <GuideRouterWithErrorAndSpinner
            apiState={guideData.state}
            ws={ws}
            guideData={guideData.data}
            getGuideData={getGuideData}
        />
    );
}

GuideRouter.propTypes = {
    ws: PropTypes.object.isRequired,
    guideData: PropTypes.object.isRequired,
    getGuideData: PropTypes.func.isRequired,
};

function GuideRouter({ ws, guideData, getGuideData }) {
    const {
        available_workflows, workflow_steps, next_step, passed, current_workflow,
    } = guideData;

    return (
        <BrowserRouter basename={`${REFORIS_URL_PREFIX}${GUIDE_URL_PREFIX}`}>
            <AlertContextProvider>
                <Portal containerId="guide-nav-container">
                    <GuideNavigation
                        workflow_steps={workflow_steps}
                        passed={passed}
                        next_step={next_step}
                    />
                </Portal>
                <Switch>
                    {workflow_steps.map((step) => {
                        const Component = STEPS[step].component;
                        return (
                            <Route
                                exact
                                key={step}
                                path={`/${step}`}
                                render={() => (
                                    <>
                                        <GuideHelp
                                            step={step}
                                            workflow={current_workflow}
                                            completed={passed.includes(step)}
                                        />
                                        <Component
                                            ws={ws}
                                            next_step={next_step}
                                            postCallback={getGuideData}
                                            workflows={available_workflows}
                                        />
                                    </>
                                )}
                            />
                        );
                    })}
                    <Redirect to={`/${next_step}`} />
                </Switch>
            </AlertContextProvider>
        </BrowserRouter>
    );
}

const GuideRouterWithErrorAndSpinner = withSpinnerOnSending(withErrorMessage(GuideRouter));
