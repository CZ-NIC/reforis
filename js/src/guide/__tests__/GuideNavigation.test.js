/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { getByText, render, wait } from "foris/testUtils/customTestRender";

import GuideNavigation from "../GuideNavigation/GuideNavigation";
import { guideFixtures } from "./__fixtures__/guide";

describe("<GuideNavigation/>", () => {
    let guideNavigationContainer;

    beforeEach(async () => {
        const { container } = render(<GuideNavigation {...guideFixtures} />);
        await wait(() => getByText(container, /Password/));
        guideNavigationContainer = container;
    });

    it("Snapshot.", () => {
        expect(guideNavigationContainer).toMatchSnapshot();
    });
});
