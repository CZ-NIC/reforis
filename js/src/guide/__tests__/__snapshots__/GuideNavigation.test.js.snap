// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GuideNavigation/> Snapshot. 1`] = `
<div>
  <ul
    class="list-unstyled"
  >
    <li>
      <a
        class="passed "
        href="/password"
      >
        <i
          class="fas fa-arrow-right"
        />
         
        Password
      </a>
    </li>
    <li>
      <a
        class="passed "
        href="/profile"
      >
        <i
          class="fas fa-arrow-right"
        />
         
        Workflow
      </a>
    </li>
    <li>
      <a
        class=" next"
        href="/networks"
      >
        <i
          class="fas fa-arrow-right"
        />
         
        Interfaces
      </a>
    </li>
    <li>
      <a>
        <i
          class="fas fa-arrow-right"
        />
         
        WAN
      </a>
    </li>
    <li>
      <a>
        <i
          class="fas fa-arrow-right"
        />
         
        Time
      </a>
    </li>
    <li>
      <a>
        <i
          class="fas fa-arrow-right"
        />
         
        DNS
      </a>
    </li>
    <li>
      <a>
        <i
          class="fas fa-arrow-right"
        />
         
        Updates
      </a>
    </li>
    <li>
      <a>
        <i
          class="fas fa-arrow-right"
        />
         
        Finish
      </a>
    </li>
  </ul>
  <a
    class="btn btn-lg btn-light "
    href="/networks"
    id="next-step-button"
  >
    <i
      class="fas fa-arrow-right"
    />
  </a>
  <button
    class="btn btn-link"
    type="button"
  >
    Skip guide
  </button>
</div>
`;
